src/main/java/de/tum/in/www1/artemis/service/LearningPathService.java

```diff
@@ -0,0 +1,411 @@
+package de.tum.in.www1.artemis.service;
+
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.stream.Collectors;
+import java.util.stream.LongStream;
+import java.util.stream.Stream;
+
+import javax.validation.constraints.NotNull;
+
+import org.jgrapht.alg.util.UnionFind;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.data.domain.Page;
+import org.springframework.stereotype.Service;
+
+import de.tum.in.www1.artemis.domain.Course;
+import de.tum.in.www1.artemis.domain.LearningObject;
+import de.tum.in.www1.artemis.domain.User;
+import de.tum.in.www1.artemis.domain.competency.Competency;
+import de.tum.in.www1.artemis.domain.competency.CompetencyRelation;
+import de.tum.in.www1.artemis.domain.competency.LearningPath;
+import de.tum.in.www1.artemis.repository.*;
+import de.tum.in.www1.artemis.web.rest.dto.PageableSearchDTO;
+import de.tum.in.www1.artemis.web.rest.dto.SearchResultPageDTO;
+import de.tum.in.www1.artemis.web.rest.dto.learningpath.NgxLearningPathDTO;
+import de.tum.in.www1.artemis.web.rest.util.PageUtil;
+
+@Service
+public class LearningPathService {
+
+    private final Logger log = LoggerFactory.getLogger(LearningPathService.class);
+
+    private final UserRepository userRepository;
+
+    private final LearningPathRepository learningPathRepository;
+
+    private final CompetencyProgressRepository competencyProgressRepository;
+
+    private final CourseRepository courseRepository;
+
+    private final CompetencyRelationRepository competencyRelationRepository;
+
+    private final LectureUnitRepository lectureUnitRepository;
+
+    private final ExerciseRepository exerciseRepository;
+
+    public LearningPathService(UserRepository userRepository, LearningPathRepository learningPathRepository, CompetencyProgressRepository competencyProgressRepository,
+            CourseRepository courseRepository, CompetencyRelationRepository competencyRelationRepository, LectureUnitRepository lectureUnitRepository,
+            ExerciseRepository exerciseRepository) {
+        this.userRepository = userRepository;
+        this.learningPathRepository = learningPathRepository;
+        this.competencyProgressRepository = competencyProgressRepository;
+        this.courseRepository = courseRepository;
+        this.competencyRelationRepository = competencyRelationRepository;
+        this.lectureUnitRepository = lectureUnitRepository;
+        this.exerciseRepository = exerciseRepository;
+    }
+
+    /**
+     * Generate learning paths for all students enrolled in the course
+     *
+     * @param course course the learning paths are created for
+     */
+    public void generateLearningPaths(@NotNull Course course) {
```

Comment:
You sometimes have `@NotNull` and sometimes `final` in front of your parameters. Is there any system behind that?

Reply:
Now all parameters of public methods are documented `@NotNull` for other developers. For private methods, this seems not necessary.
I removed all `final` parameters, due to our SCA this is not necessary anymore. Thx for pointing this out 👍